server/src/main/java/org/elasticsearch/transport/TransportHandshaker.java

```diff
@@ -149,24 +161,78 @@ public String executor() {
         }
     }
 
-    static final class VersionHandshakeResponse extends TransportResponse {
+    static final class HandshakeRequest extends TransportRequest {
 
         private final Version version;
 
-        VersionHandshakeResponse(Version version) {
+        HandshakeRequest(Version version) {
             this.version = version;
         }
 
-        private VersionHandshakeResponse(StreamInput in) throws IOException {
+        HandshakeRequest(StreamInput streamInput) throws IOException {
+            super(streamInput);
+            int remainingMessageBytes;
+            try {
+                remainingMessageBytes = streamInput.readInt();
```

Comment:
can we use `StreamInput#readBytesReference()` here. It uses `readArraySize` which adds additional safety. But then we also need to use `writeBytesReference` on the write end.

Reply:
I made this change.